### PR TITLE
Unpin Sphinx

### PR DESCRIPTION
**Pull request by saimn**
Ref #11725 and https://github.com/astropy/sphinx-automodapi/pull/150

So basically using `:noindex:` with https://github.com/astropy/sphinx-automodapi/pull/150 works in the sense that it removes the duplication warnings from Sphinx. But then for Astropy we get other warnings/errors about missing references ...

### CODE REVIEW

**Comment by saimn** [comment]
Example of the new warnings with this PR:
```
.../astropy/coordinates/builtin_frames/hadec.py:docstring of astropy.coordinates.builtin_frames.hadec.HADec:1: WARNING: py:class reference target not found: astropy.coordinates.baseframe.BaseCoordinateFrame
.../astropy/coordinates/builtin_frames/hcrs.py:docstring of astropy.coordinates.builtin_frames.hcrs.HCRS:1: WARNING: py:class reference target not found: astropy.coordinates.builtin_frames.baseradec.BaseRADecFrame
.../astropy/coordinates/builtin_frames/ecliptic.py:docstring of astropy.coordinates.builtin_frames.ecliptic.HeliocentricEclipticIAU76:1: WARNING: py:class reference target not found: astropy.coordinates.builtin_frames.ecliptic.BaseEclipticFrame
.../astropy/coordinates/builtin_frames/ecliptic.py:docstring of astropy.coordinates.builtin_frames.ecliptic.HeliocentricMeanEcliptic:1: WARNING: py:class reference target not found: astropy.coordinates.builtin_frames.ecliptic.BaseEclipticFrame
.../astropy/coordinates/builtin_frames/ecliptic.py:docstring of astropy.coordinates.builtin_frames.ecliptic.HeliocentricTrueEcliptic:1: WARNING: py:class reference target not found: astropy.coordinates.builtin_frames.ecliptic.BaseEclipticFrame
```

**Comment by pllim** [comment]
@adrn , can we simplify `coordinates`? 😆

**Comment by adrn** [comment]
Hm, for the warnings in coordinates, is this because the `Base*Frame` API pages were being generated by the index made from automodapi'ing `astropy.coordinates.builtin_frames`?

**Comment by saimn** [comment]
See #14115.

**Comment by mhvk** [comment]
@saimn - can you re-open this? I fixed `astropy.units` at least which should give an idea about how to proceed... I tried pushing to your branch and suspect that failed because the PR is closed.

**Comment by pllim** [comment]
I reopened it though this might need a rebase by now...

**Comment by saimn** [comment]
I thought this PR was not needed/useful, do we still need `:noindex:` if we change the way to document those subpackages ?

**Comment by mhvk** [comment]
I'm using `:noindex:` in `astropy.units.function`. Perhaps we can indeed just not show it, but maybe it is good to just test that it works?

**Comment by mhvk** [comment]
I think I may have it down to one that I don't quite understand, but we'll see.  Also, need to check whether `:noindex:` is indeed necessary, but time to start preparing some dinner now!

**Comment by mhvk** [comment]
OK, I was wrong -- still lots in `coordinates`. I think it does need a `:noindex:` in `builtin_frames` after all (somehow the local built seemed better). But at least some progress!

File: docs/coordinates/index.rst [hunk 1]
```diff
@@ -566,3 +566,35 @@ Reference/API
 =============
 
 .. automodapi:: astropy.coordinates
+    :skip: ICRS
```
Comment (Cadair) [inline docs/coordinates/index.rst:569]: This obviously isn't great (it's the whole contents of the `__all__`). I don't think it substantially detracts from the documentation of this page (although maybe it needs a note to say things aren't listed here?). Ideally of course we would have a way of excluding objects on namespace, i.e. do something like `:exclude-namespace: astropy.coordinates.builtin_frames` but :man_shrugging:

File: astropy/coordinates/builtin_frames/galactocentric.py [hunk 2]
```diff
@@ -509,11 +510,12 @@ def __init__(self, *args, **kwargs):
 
     @classmethod
     def get_roll0(cls):
```
Comment (pllim) [inline astropy/coordinates/builtin_frames/galactocentric.py:512]: Unrelated... but we might need a `get_roll20` 🤓

File: docs/conf.py [hunk 3]
```diff
@@ -335,7 +335,9 @@
         "reference_url": {
             "astropy": None,
             "matplotlib": "https://matplotlib.org/stable/",
-            "numpy": "https://numpy.org/doc/stable/",
+            # The numpy search js isn't loadable at the moment (2022-12-07)
```
Comment (pllim) [inline docs/conf.py:338]: Wait, what?

File: docs/coordinates/index.rst (hunk 1 above)
Comment (pllim) [reply docs/coordinates/index.rst:569]: As long as we add a line that says "go here for frame classes" or something, I think this is acceptable if it is the only way we can unpin Sphinx max version.

File: setup.cfg [hunk 4]
```diff
@@ -97,8 +97,9 @@ all =
     fsspec[http]>=2022.8.2
     s3fs>=2022.8.2
 docs =
-    sphinx<4
+    sphinx
```
Comment (pllim) [inline setup.cfg:100]: Also change the Jinja2 pin back to this:

```
Jinja2>=3.0
```

**Comment by mhvk** [comment]
Nice!!!  I think now you narrowed it down we should be able to get rid of the problems with the graph. It seems like we do not really need `:noindex:` -- I think we can simply omit `astropy.units.function` (though will think a bit more about that).

**Comment by pllim** [comment]
Given all the work Simon has poured into making noindex work, I feel like we are obligated to use it. 😆

File: docs/coordinates/index.rst (hunk 1 above)
Comment (Cadair) [reply docs/coordinates/index.rst:569]: The classes are listed in the block above, so it's not that hard to spot but I am happy to add a line.

**Review by larrybradley — COMMENTED** [review]
Note that  Sphinx 5.2 introduced a new feature (that is on by default) with a table of contents for functions and classes.  That text overflows the left side bar window in the astropy sphinx theme.  See, e.g., from this PR: https://astropy--14093.org.readthedocs.build/en/14093/api/astropy.table.Table.html#astropy.table.Table

Fortunately, Sphinx 5.2.3 added a configuration option to turn it off:  set `toc_object_entries = False` in `docs.conf`.

xref: https://github.com/astropy/photutils/pull/1424
xref: https://github.com/astropy/photutils/pull/1465/files

File: docs/conf.py (hunk 3 above)
Comment (Cadair) [reply docs/conf.py:338]: This was causing an error, and it's because the JS is missing some quotes apparently :man_shrugging:

Comment (pllim) [reply docs/conf.py:338]: But I don't see any such failure in other PRs. Is the JS only used for this PR?

File: docs/coordinates/index.rst (hunk 1 above)
Comment (mhvk) [reply docs/coordinates/index.rst:569]: Maybe good to see if we can change the transformgraph view to not link to classes - that seems to be the problem (and even now we cannot click on it anyway)

Comment (Cadair) [reply docs/coordinates/index.rst:569]: Is there actually a problem, other than the links inside the graph not working? (I am not sure they did to begin with)

**Comment by mhvk** [comment]
I think I've got an alternative without `:noindex:` by using `automodule` instead of `automodapi`. Maybe easiest to make a separate PR so it is easier to compare.

**Comment by pllim** [comment]
I pushed a commit to address https://github.com/astropy/astropy/pull/14093#pullrequestreview-1208664859 . Thanks, @larrybradley !

**Comment by Cadair** [comment]
Closed in favour of #14140

**Pull request closed without merging**